common/src/main/java/net/opentsdb/threadpools/UserAwareThreadPoolExecutor.java

```diff
@@ -0,0 +1,352 @@
+// This file is part of OpenTSDB.
+// Copyright (C) 2018 The OpenTSDB Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package net.opentsdb.threadpools;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.RunnableFuture;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.stumbleupon.async.Deferred;
+
+import net.opentsdb.core.TSDB;
+import net.opentsdb.query.QueryContext;
+
+public class UserAwareThreadPoolExecutor implements TSDBThreadPoolExecutor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(UserAwareThreadPoolExecutor.class);
+
+  public static final String TYPE = "UserAwareThreadPoolExecutor";
+
+  /** Percentage of active threads to consider before applying the reschedule */
+  protected static final String THREAD_THRESHOLD_PCT = "thread.threshold.pct";
+
+  /**
+   * Percentage of active threads per user to consider before applying the reschedule when the total
+   * percentage of active thread in the executor falls below this threshold value
+   */
+  protected static final String MAX_THREAD_PER_USER_PCT = "max.thread.per.user.pct";
+
+  /** Stores the User thread count that's being current executed */
+  private static final ConcurrentHashMap<String, AtomicInteger> CURRENT_EXECUTIONS =
```

Comment:
No statics please. We can just have it as a member of the plugin.

Reply:
rgr